cratedb_toolkit/wtf/library.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) 2021-2023, Crate.io Inc.
+# Distributed under the terms of the AGPLv3 license, see LICENSE.
+
+
+class Library:
+    """
+    A collection of SQL queries and utilities suitable for diagnostics on CrateDB.
+    """
+
+    class Health:
+        """
+        CrateDB health check queries.
+
+        References:
+
+        - https://community.cratedb.com/t/similar-elasticsearch-commands/1455/4
+          Thanks, Eduardo Legatti.
+        """
+
+        cluster_name = """
+        SELECT name FROM sys.cluster;
+        """
+
+        table_allocations = """
+        SELECT table_schema, table_name, node_id, shard_id, partition_ident, current_state, decisions, explanation
+        FROM sys.allocations;
+        """
+
+        table_allocations_special = """
+        SELECT decisions[2]['node_name'] AS node_name, COUNT(*) AS table_count
+        FROM sys.allocations
+        GROUP BY decisions[2]['node_name'];
+        """
+
+        table_health = """
+        SELECT health, COUNT(*) AS table_count FROM sys.health GROUP BY health;
+        """
+
+        nodes_count = """
+        SELECT COUNT(id) AS count FROM sys.nodes;
+        """
+        nodes_list = """
+        SELECT * FROM sys.nodes ORDER BY hostname;
+        """
+
+    class Logs:
+        # TODO: Implement `tail` in one way or another. -- https://stackoverflow.com/q/4714975
+        #       SELECT * FROM sys.jobs_log OFFSET -10;
+        #       SELECT * FROM sys.jobs_log OFFSET (SELECT count(*) FROM sys.jobs_log)-10;
+        # https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#to-char-expression-format-string
+        # https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#date-format-format-string-timezone-timestamp
+        user_queries = """
+        SELECT
+            DATE_FORMAT('%Y-%m-%dT%H:%i:%s.%f', started) AS started,
+            DATE_FORMAT('%Y-%m-%dT%H:%i:%s.%f', ended) AS ended,
+            classification, stmt, username, node
+        FROM
+            sys.jobs_log
+        WHERE
+            stmt NOT LIKE '%sys.%' AND
+            stmt NOT LIKE '%information_schema.%'
+        """
```

Comment:
Dear @hlcianfagna and @hammerhead,

I would like to implement tailing the `sys.jobs_log` table in one way or another. Actually, `tail --follow`. Did you manage to that yet?

Cheers,
Andreas.

Reply:
```sql
 SELECT CURRENT_TIMESTAMP AS last_timestamp,
                            (ended / 10000) * 10000 + 5000 AS ended_time,
                            COUNT(*) / 10.0 AS qps,
                            TRUNC(AVG(ended::bigint - started::bigint), 2) AS duration,
                            UPPER(regexp_matches(stmt, '^\s*(\w+).*') [1]) AS query_type
FROM sys.jobs_log
WHERE ended > now() - ('15 minutes')::interval
GROUP BY 1,
         2,
         5
ORDER BY ended_time ASC
```

This is the query used in the panel to show the latest 15 minutes QPS, perhaps it can serve as inspiration :P

Reply:
It seems empty when it doesn't have a value.

Something like
##### Query ran at 2023-12-02T11:02:47.288Z on CrateDB 5.5.0
```sql
SELECT
  DATE_FORMAT(started)
FROM
  "sys"."jobs_log"
ORDER BY
  started DESC
LIMIT
  10
```
##### QUERY OK, 10 record(s) returned in 0.0045s

|date_format(started)|
 |-| 
|"2023-12-02T11:02:40.890000Z"|
|"2023-12-02T11:02:40.825000Z"|
|"2023-12-02T11:02:40.822000Z"|
|"2023-12-02T11:02:40.819000Z"|
|"2023-12-02T11:02:40.819000Z"|
|"2023-12-02T11:02:40.819000Z"|
|"2023-12-02T11:02:40.819000Z"|
|"2023-12-02T11:02:35.892000Z"|
|"2023-12-02T11:02:35.821000Z"|
|"2023-12-02T11:02:35.821000Z"|

Seem to work just fine as `tail -f`

Reply:
To follow what is new I would keep track of the last started timestamp observed and query what is new

Reply:
In order to get this patch merged, we postponed this by adding it as an item to the backlog.

https://github.com/crate-workbench/cratedb-toolkit/blob/1c437cad2533c2e47423e7c32a6f15e79f5adcd1/doc/wtf/backlog.md?plain=1#L5-L6

Reply:
We created a dedicated issue, because it's a prominent feature yet missing.

- GH-171